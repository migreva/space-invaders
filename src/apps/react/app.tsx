import { GameCanvas } from '@space-invaders/apps/react/components/game-canvas/game-canvas';
import { StoreState } from '@space-invaders/store';
import * as React from 'react';

interface AppProps {
	state: StoreState;
}

export function App(props: AppProps): JSX.Element {
	return (
		<div className="app-container">
				<GameCanvas {...props}/>
		</div>
	);
}
