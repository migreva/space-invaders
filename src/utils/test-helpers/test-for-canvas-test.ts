import { CanvasRenderingContext2DEvent } from 'jest-canvas-mock';

const FILL_TEXT_CANVAS_EVENT_TYPE: string = 'fillText';

function checkForCanvasText(canvas: HTMLCanvasElement, text: string): CanvasRenderingContext2DEvent[] { 
	const ctx: CanvasRenderingContext2D = canvas.getContext('2d');
	const events: CanvasRenderingContext2DEvent[] = ctx.__getEvents();
	// console.log(events);

	// filter all canvas events down, and ensure a `fillText` event 
	// that matches the text exists
	return events
		.filter((canvasEvent) => canvasEvent.type === FILL_TEXT_CANVAS_EVENT_TYPE)
		.filter((canvasEvent) => canvasEvent.props.text?.toLowerCase().indexOf(text) !== 0);
}

/**
 * assert that certain text exists in the canvas
 * 
 * @param canvas the canvas to check
 * @param text the text to assert exists in the canvas
 */
export function assertCanvasTextExists(canvas: HTMLCanvasElement, text: string) {
	// we should have at least one fillText event that matches
	expect(checkForCanvasText(canvas, text).length >= 1).toBe(true);
}

/**
 * assert that certain text exists in the canvas
 * 
 * @param canvas the canvas to check
 * @param text the text to assert exists in the canvas
 */
export function assertCanvasTextDoesNotExists(canvas: HTMLCanvasElement, text: string) {
	// we should have at least one fillText event that matches
	expect(checkForCanvasText(canvas, text)).toHaveLength(0);
}


