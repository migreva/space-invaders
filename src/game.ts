
import { App } from '@space-invaders/apps/react/app';
import { store } from '@space-invaders/store';
import { handleUserInput, startGameTick, stopGameTick } from '@space-invaders/store/state';
import { createRoot, Root } from 'react-dom/client';
import { AnyAction } from 'redux';

/**
 * render a react app
 * 
 * @param root React root
 */
function renderApp(root: Root): void {
	// render our react app
	root.render(App({
		state: store.getState(),
	}));
}

/**
 * 
 * @param e keydown 
 */
function userKeydownInput(e: KeyboardEvent): void {
	store.dispatch(handleUserInput(e) as unknown as AnyAction);
}

export function startGame(): void {
	// init the root HTML element
	const rootHTMLElement: HTMLDivElement = document.createElement('div');
	document.body.appendChild(rootHTMLElement);

	// TODO segment this out to init the correct app based on URL
	// create app, subscribe to event state
	const root: Root = createRoot(rootHTMLElement);
	renderApp(root);
	store.subscribe(() => {
		renderApp(root);
	});

	// start the tick
	store.dispatch(startGameTick() as unknown as AnyAction);

	// track user keydown events
	document.addEventListener('keydown', userKeydownInput);
}

export function stopGame() {
	store.dispatch(stopGameTick() as unknown as AnyAction);
	document.removeEventListener('keydown', userKeydownInput);
}
