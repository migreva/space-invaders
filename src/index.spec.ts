/**
 * @jest-environment jsdom
 */

import { startGame, stopGame } from '@space-invaders/game';
import { assertCanvasTextDoesNotExists, assertCanvasTextExists } from '@space-invaders/utils/test-helpers/test-for-canvas-test';
import '@testing-library/jest-dom';
import { fireEvent } from '@testing-library/dom';
import { screen } from '@testing-library/react';
import 'jest-canvas-mock';
import { act } from 'react-dom/test-utils';

beforeEach(() => {
	act(() => {
		// start a game
		startGame();

		// run a single game tick
		jest.runOnlyPendingTimers();
	});
});

afterEach(() => {
	act(() => stopGame());

	// https://stackoverflow.com/a/50800473
	// to clear out the game in between tests
	document.getElementsByTagName('html')[0].innerHTML = ''; 
});

test('should render a canvas element', async (): Promise<void> => {
	await screen.findByTestId('game-canvas');
});

test('should pause the game when escape is pressed', async (): Promise<void> => {
	// first, ensure "Paused" does not exist
	const PAUSE_TEXT: string = 'Paused';
	const canvas: HTMLCanvasElement = await screen.findByTestId('game-canvas');
	assertCanvasTextDoesNotExists(
		canvas, 
		PAUSE_TEXT,
	);

	// pause the game and ensure the text exists
	fireEvent.keyDown(document, {
		key: 'Escape',
		code: 'Escape',
		keyCode: 192,
	});
	assertCanvasTextExists(
		canvas, 
		PAUSE_TEXT,
	);

	// now un-pause the game by keystroke
	canvas.getContext('2d').__clearEvents();
	fireEvent.keyDown(document, {
		key: 'Escape',
		code: 'Escape',
		keyCode: 192,
	});
	assertCanvasTextDoesNotExists(
		canvas, 
		PAUSE_TEXT,
	);
});


// test('should handle basic movement', async (): Promise<void> => {
	// const canvas: HTMLCanvasElement = await screen.findByTestId('game-canvas');
	// const ctx: CanvasRenderingContext2D = canvas.getContext('2d');
	// const events = ctx.__getEvents();
	// console.log(events);

	// expect(events).toMatchSnapshot();
// });
