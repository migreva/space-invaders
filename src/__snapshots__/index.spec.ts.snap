// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle basic movement 1`] = `
[
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "maxWidth": null,
      "text": "Paused (hit Esc to unpause)",
      "x": 0,
      "y": 10,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "maxWidth": null,
      "text": "Paused (hit Esc to unpause)",
      "x": 0,
      "y": 10,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "height": 100,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "height": 10,
      "width": 10,
      "x": 100,
      "y": 90,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;
